powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/service/CredentialCounterService.java

```diff
@@ -124,9 +123,69 @@ public void updateCredentialCounter(CredentialEntity credential, AuthenticationR
-        logger.info("Credential counter updated, user ID: {}, credential definition name: {}, attempt counter: {}, soft counter: {}, hard counter: {}, status: {}",
-                credential.getUser().getUserId(), credential.getCredentialDefinition().getName(),
+        return credentialRepository.save(credential);
+    }
+
+    private CredentialEntity updateCounterLdap(CredentialEntity credential, AuthenticationResult authenticationResult) throws InvalidRequestException {
+
+        switch (authenticationResult) {
+            case SUCCEEDED -> {
+                credential.setFailedAttemptCounterSoft(0);
+                credential.setFailedAttemptCounterHard(0);
+            }
+            case FAILED -> {
+                ExternalCredentialDetail externalCredentialDetail = ldapVerifierService.readCredentialExternalStatus(credential);
+                credential.setFailedAttemptCounterSoft(externalCredentialDetail.getFailedAttemptsOpt().orElse(0));
+                credential.setFailedAttemptCounterHard(externalCredentialDetail.getFailedAttemptsOpt().orElse(0));
+            }
+            default -> throw new InvalidRequestException("Invalid authentication result: " + authenticationResult);
+        }
+        return credentialRepository.save(credential);
+    }
+
+    private CredentialEntity updateCounterProxy(CredentialEntity credential, AuthenticationResult authenticationResult) throws InvalidRequestException {
+
+        switch (authenticationResult) {
+            case SUCCEEDED -> {
+                credential.setFailedAttemptCounterSoft(0);
+                credential.setFailedAttemptCounterHard(0);
+            }
+            case FAILED -> {
+                credential.setFailedAttemptCounterSoft(credential.getFailedAttemptCounterSoft() + 1);
+                credential.setFailedAttemptCounterHard(credential.getFailedAttemptCounterHard() + 1);
+            }
+            default -> throw new InvalidRequestException("Invalid authentication result: " + authenticationResult);
+        }
+        return credentialRepository.save(credential);
+
+    }
+
+    /**
+     * Update credential counter. This method is not transactional.
+     * @param credential Credential entity.
+     * @param authenticationResult Authentication result.
+     * @throws InvalidRequestException Thrown when request is invalid.
+     */
+    public void updateCredentialCounter(CredentialEntity credential, AuthenticationResult authenticationResult) throws InvalidRequestException {
+        logger.info("action: {}, state: {} userId: {}, getCredentialDefinition: {}, authenticationResult: {}",
+                "updateCredentialCounter", "initiated",
+                credential.getUser().getUserId(), credential.getCredentialDefinition().getName(), authenticationResult);
+        switch (credential.getSource()) {
+            case LOCAL -> {
+                credential = updateCounterInternal(credential, authenticationResult);
+            }
```

Comment:
Labeled rule's code block is redundant
```suggestion
            case LOCAL -> credential = updateCounterInternal(credential, authenticationResult);
```

Reply:
Nitpick: I still see the unnecessary curly braces.